scoreboard.pl

```diff
@@ -607,224 +714,81 @@ ($)
-                push @existing, "$cache_dir/$1";
-            }
-        }
-        undef $dh;
-
         # Find any new logs to make pretty
         # We do this in oldest to newest order since the
         # Prettify Process will update the latest.txt file
-        @existing = sort @existing;
-        my $updated = 0;
+        my @existing;
+        if (!list_logs(\@existing, $build_dir)) {
+            next;
+        }
+        my $keep = (defined $builds{$buildname}->{KEEP}) ? $builds{$buildname}->{KEEP} : $keep_default;
+        my $updated = delete_old_logs(\@existing, $keep);
 
         # A trigger file to tells the scoreboard that the log is complete
         # The reason for this that that there is a race condition where the
         # .txt file exist but not be competely copied.  There is a new
         # copy subcommand to process_logs that creates this trigger file
         # after the copy is complete, but doing it here works even if
         # the autobuild uses move, just delayed one iteration.
-        my $triggerfile = "$directory/$buildname/post";
-        my $post = 0;
-        if ( -e $triggerfile ) {
-            $post = 1;
-            unlink $triggerfile;
-        }
-        print "        in local_update_cache, post=$post\n" if $verbose;
-
-        # Get info from the latest build
-        my $file_name1 = "$directory/$buildname/latest.txt";
-        my $file_handle1 = new FileHandle ($file_name1, 'r');
-        my $latest_basename = "";
-        if (defined $file_handle1) {
-            while (<$file_handle1>) {
-                if ($_ =~ m/(...._.._.._.._..) /) {
-                     $latest_basename = $1;
-                }
-            }
-        }
-        undef $file_handle1;
+        my $post = check_post($build_dir);
+        print "        post=$post, use_build_logs=$use_build_logs\n" if ($verbose);
 
+        my $latest = get_latest($build_dir);
         foreach my $file (@existing) {
-            if ( -e $file . "_Totals.html" || $post == 1 ) {
-                # skip scenario when Failed Test Log is not needed, and all other logs already exist
-                if (!($use_build_logs && (-e $file . "_Totals.html"))) {
-                    # process only the latest text file if logs already exist
-                    next if ((-e $file . "_Totals.html") && !($latest_basename eq substr($file, -length($latest_basename))));
-                    print "        Prettifying $file.txt\n" if($verbose);
-                    Prettify::Process ("$file.txt", $buildname, $failed_tests_by_test_ref, $use_build_logs, $builds{$buildname}->{DIFFROOT}, "$directory/$log_prefix", (-e $file . "_Totals.html"));
-                    $updated++;
-                }
-            } else {
-                # Create the triggerfile for the next time we run
-                open(FH, ">$triggerfile");
-                close(FH);
-                last;
+            my $totals_exist = (-e "${file}_Totals.html") && !(-z _);
```

Comment:
What is "_" here? It seems like this is a mistake, but I don't know perl well enough to be sure.

Reply:
It's the same file that the previous operator used.
See http://www.rexswain.com/perl5.html under File Test Operators

Reply:
That's what I was wondering. In that case, this seems OK.

Reply:
"Whenever you use the file test operator, Perl will make a new call of stat() , which can be expensive. However, Perl stores the result from the last stat() call to a special filehandle named _ , so the subsequent file test operators can use the result that stores in the _ filehandle."